# Data Cleaning

In [1]:
import sys, os
sys.path.append('../')
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from utils.cleaning import cleaning_pipeline, DataFrameAnalyzer
import warnings
warnings.filterwarnings("ignore")

# Data Ingestion

In [2]:
df_previous = pd.read_csv("./Data/air_system_previous_years.csv")
print(f"Shape: {df_previous.shape}")
df_previous.head()

Shape: (60000, 171)


class  aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  \
0   neg   76698     na  2130706438    280      0      0      0      0      0   
1   neg   33058     na           0     na      0      0      0      0      0   
2   neg   41040     na         228    100      0      0      0      0      0   
3   neg      12      0          70     66      0     10      0      0      0   
4   neg   60874     na        1368    458      0      0      0      0      0   

   ...   ee_002  ee_003  ee_004  ee_005  ee_006  ee_007  ee_008 ee_009 ef_000  \
0  ...  1240520  493384  721044  469792  339156  157956   73224      0      0   
1  ...   421400  178064  293306  245416  133654   81140   97576   1500      0   
2  ...   277378  159812  423992  409564  320746  158022   95128    514      0   
3  ...      240      46      58      44      10       0       0      0      4   
4  ...   622012  229790  405298  347188  286954  311560  433954   1218      0   

  eg_000  
0      0  
1      0  
2      0  
3     32  
4      0  

[5 rows x 171 columns]

In [62]:
df_present = pd.read_csv("./Data/air_system_present_year.csv")
print(f"Shape: {df_present.shape}")
df_present.head()

Shape: (16000, 171)


class  aa_000 ab_000 ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  ...  \
0   neg      60      0     20     12      0      0      0      0      0  ...   
1   neg      82      0     68     40      0      0      0      0      0  ...   
2   neg   66002      2    212    112      0      0      0      0      0  ...   
3   neg   59816     na   1010    936      0      0      0      0      0  ...   
4   neg    1814     na    156    140      0      0      0      0      0  ...   

   ee_002  ee_003  ee_004  ee_005   ee_006  ee_007  ee_008 ee_009 ef_000  \
0    1098     138     412     654       78      88       0      0      0   
1    1068     276    1620     116       86     462       0      0      0   
2  495076  380368  440134  269556  1315022  153680     516      0      0   
3  540820  243270  483302  485332   431376  210074  281662   3232      0   
4    7646    4144   18466   49782     3176     482      76      0      0   

  eg_000  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 171 columns]

In [63]:
analyzer_previous = DataFrameAnalyzer(df_previous)
analyzer_present = DataFrameAnalyzer(df_present)

# General Look at the Data

## Previous Years

In [64]:
df_previous.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60000 entries, 0 to 59999
Columns: 171 entries, class to eg_000
dtypes: int64(1), object(170)
memory usage: 78.3+ MB


In [65]:
na_columns_previous = analyzer_previous.find_na_columns()
na_columns_previous

{'ab_000': {'na_count': 46329, 'dtype': dtype('O')},
 'ac_000': {'na_count': 3335, 'dtype': dtype('O')},
 'ad_000': {'na_count': 14861, 'dtype': dtype('O')},
 'ae_000': {'na_count': 2500, 'dtype': dtype('O')},
 'af_000': {'na_count': 2500, 'dtype': dtype('O')},
 'ag_000': {'na_count': 671, 'dtype': dtype('O')},
 'ag_001': {'na_count': 671, 'dtype': dtype('O')},
 'ag_002': {'na_count': 671, 'dtype': dtype('O')},
 'ag_003': {'na_count': 671, 'dtype': dtype('O')},
 'ag_004': {'na_count': 671, 'dtype': dtype('O')},
 'ag_005': {'na_count': 671, 'dtype': dtype('O')},
 'ag_006': {'na_count': 671, 'dtype': dtype('O')},
 'ag_007': {'na_count': 671, 'dtype': dtype('O')},
 'ag_008': {'na_count': 671, 'dtype': dtype('O')},
 'ag_009': {'na_count': 671, 'dtype': dtype('O')},
 'ah_000': {'na_count': 645, 'dtype': dtype('O')},
 'ai_000': {'na_count': 629, 'dtype': dtype('O')},
 'aj_000': {'na_count': 629, 'dtype': dtype('O')},
 'ak_000': {'na_count': 4400, 'dtype': dtype('O')},
 'al_000': {'na_count':

In [66]:
non_na_columns = set(df_previous.columns.to_list()) - set(list(na_columns_previous.keys()))
print(f"Columns without 'na' values: {non_na_columns}")

Columns without 'na' values: {'aa_000', 'class'}


In [67]:
df_previous.duplicated().sum()

0

In [68]:
columns_with_non_digit_non_na = [col for col in df_previous.columns if analyzer_previous.check_non_digit_and_non_na(col)]
print(f"Columns with values that are not digits, floats, and not 'na': {columns_with_non_digit_non_na}")

Columns with values that are not digits, floats, and not 'na': ['class']


## Present Year

In [69]:
df_present.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16000 entries, 0 to 15999
Columns: 171 entries, class to eg_000
dtypes: int64(1), object(170)
memory usage: 20.9+ MB


In [70]:
na_columns_present = analyzer_present.find_na_columns()
na_columns_present

{'ab_000': {'na_count': 12363, 'dtype': dtype('O')},
 'ac_000': {'na_count': 926, 'dtype': dtype('O')},
 'ad_000': {'na_count': 3981, 'dtype': dtype('O')},
 'ae_000': {'na_count': 690, 'dtype': dtype('O')},
 'af_000': {'na_count': 690, 'dtype': dtype('O')},
 'ag_000': {'na_count': 189, 'dtype': dtype('O')},
 'ag_001': {'na_count': 189, 'dtype': dtype('O')},
 'ag_002': {'na_count': 189, 'dtype': dtype('O')},
 'ag_003': {'na_count': 189, 'dtype': dtype('O')},
 'ag_004': {'na_count': 189, 'dtype': dtype('O')},
 'ag_005': {'na_count': 189, 'dtype': dtype('O')},
 'ag_006': {'na_count': 189, 'dtype': dtype('O')},
 'ag_007': {'na_count': 189, 'dtype': dtype('O')},
 'ag_008': {'na_count': 189, 'dtype': dtype('O')},
 'ag_009': {'na_count': 189, 'dtype': dtype('O')},
 'ah_000': {'na_count': 175, 'dtype': dtype('O')},
 'ai_000': {'na_count': 163, 'dtype': dtype('O')},
 'aj_000': {'na_count': 163, 'dtype': dtype('O')},
 'ak_000': {'na_count': 1198, 'dtype': dtype('O')},
 'al_000': {'na_count': 169

In [71]:
df_present.duplicated().sum()

0

In [72]:
columns_with_non_digit_non_na = [col for col in df_present.columns if analyzer_present.check_non_digit_and_non_na(col)]
print(f"Columns with values that are not digits, floats, and not 'na': {columns_with_non_digit_non_na}")

Columns with values that are not digits, floats, and not 'na': ['class']


### Observations:
- Both datasets behave in the exact same way
- All column names are already formatted in lowercase
- There are 171 columns in total
- Target column `class` is a string of "neg" and "pos"
- All but two columns have NaN values (represented by the string "na"). Some of these columns have 77% of their data as "na"
- There are no duplicated rows
- The only column where the values are not digits, floats or the "na" string is `class`
### Impact:
- Reducing the dimension of the dataset will be a must
- `class` needs to have its values changed from strings to numbers
- The columns can be investigated in terms of relation to the target `class`
- All columns with "na", either have values as digits or "na". This indicates that these columns can be converted to int or float, when the missing data is dealt with

# Missing Data

In [58]:
na_counts_df = pd.concat([pd.DataFrame.from_dict(na_columns_previous, orient="index", columns=["na_count"]), 
                        pd.DataFrame.from_dict(na_columns_present, orient="index", columns=["na_count"])], axis=1)
na_counts_df.columns = ["na_count_previous", "na_count_present"]
na_counts_df["na_percentage_previous"] = na_counts_df["na_count_previous"] / len(df_previous) * 100
na_counts_df["na_percentage_present"] = na_counts_df["na_count_present"] / len(df_present) * 100
na_counts_df.head()

na_count_previous  na_count_present  na_percentage_previous  \
ab_000              46329             12363               77.215000   
ac_000               3335               926                5.558333   
ad_000              14861              3981               24.768333   
ae_000               2500               690                4.166667   
af_000               2500               690                4.166667   

        na_percentage_present  
ab_000               77.26875  
ac_000                5.78750  
ad_000               24.88125  
ae_000                4.31250  
af_000                4.31250

In [59]:
columns_high_na_previous = na_counts_df[na_counts_df["na_percentage_previous"] > 50].index.tolist()
columns_high_na_present = na_counts_df[na_counts_df["na_percentage_present"] > 50].index.tolist()
columns_high_na_previous, columns_high_na_present

(['ab_000',
  'bm_000',
  'bn_000',
  'bo_000',
  'bp_000',
  'bq_000',
  'br_000',
  'cr_000'],
 ['ab_000',
  'bm_000',
  'bn_000',
  'bo_000',
  'bp_000',
  'bq_000',
  'br_000',
  'cr_000'])

### Observations:
- There are 8 columns with more than 50% of their values as "na". They are the same columns for both datasets

### Impact:
- Since more than 50% of these values are "na", these columns can be dropped
- NaN values of the remaining columns will be replaced with the median of each column, preserving the column distribution

# Data Cleaning Pipeline

### Based on the analysis in this notebook, the following data cleaning steps will be taken:
- Map the values from `class` to 0 when "neg" and 1 when "pos"
- Replace all "na" strings with np.nan
- Remove all columns where the NaN values are more than 50% of the column data
- Convert every remaining column with NaN values to int or float
- Replace all NaN values with the median of each column

In [3]:
df_clean = df_previous.copy()
cleaning_pipeline(df_clean, "Data", "clean_air_system_previous_years")